From af95df7da2663bda9c84016816f5f25248e05bdd Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Mon, 18 Aug 2014 19:58:35 -0400
Subject: [block] nvme: Improve error messages

Message-id: <1408391935-24886-3-git-send-email-dmilburn@redhat.com>
Patchwork-id: 87882
O-Subject: [RHEL7.1 PATCH BZ 1111259 02/22] NVMe: Improve error messages
Bugzilla: 1111259
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Matthew Wilcox <matthew.r.wilcox@intel.com>

Help people diagnose what is going wrong at initialisation time by
printing out which command has gone wrong and what the device returned.
Also fix the error message printed while waiting for reset.

Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Reviewed-by: Keith Busch <keith.busch@intel.com>
(cherry picked from commit 27e8166c31656f7780e682eaf6bc9c3b8dd03253)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/nvme-core.c b/drivers/block/nvme-core.c
index 5e1f8f7..0ca2368 100644
--- a/drivers/block/nvme-core.c
+++ b/drivers/block/nvme-core.c
@@ -1437,7 +1437,8 @@ static int nvme_wait_ready(struct nvme_dev *dev, u64 cap, bool enabled)
    return -EINTR;
   if (time_after(jiffies, timeout)) {
    dev_err(&dev->pci_dev->dev,
-    "Device not ready; aborting initialisation\n");
+    "Device not ready; aborting %s\n", enabled ?
+      "initialisation" : "reset");
    return -ENODEV;
   }
  }
@@ -2142,8 +2143,13 @@ static int set_queue_count(struct nvme_dev *dev, int count)
 
  status = nvme_set_features(dev, NVME_FEAT_NUM_QUEUES, q_count, 0,
         &result);
- if (status)
-  return status < 0 ? -EIO : -EBUSY;
+ if (status < 0)
+  return status;
+ if (status > 0) {
+  dev_err(&dev->pci_dev->dev, "Could not set queue count (%d)\n",
+         status);
+  return -EBUSY;
+ }
  return min(result & 0xffff, result >> 16) + 1;
 }
 
@@ -2279,6 +2285,7 @@ static int nvme_dev_add(struct nvme_dev *dev)
 
  res = nvme_identify(dev, 0, 1, dma_addr);
  if (res) {
+  dev_err(&pdev->dev, "Identify Controller failed (%d)\n", res);
   res = -EIO;
   goto out;
  }
-- 
1.7.1